src/apis/middleware.rs

```diff
@@ -0,0 +1,74 @@
+use reqwest::blocking::Client;
+use reqwest::blocking::{Request, Response};
+use reqwest::{IntoUrl, Method};
+
+use reqwest::blocking::RequestBuilder;
+
+#[derive(Debug, Clone)]
+pub struct ClientWithMiddleware {
+    inner: Client,
+    max_retries: i32,
+    retry_backoff_factor: f32,
+}
+
+impl ClientWithMiddleware {
+    pub fn new(inner: Client, max_retries: i32, retry_backoff_factor: f32) -> ClientWithMiddleware {
+        ClientWithMiddleware {
+            inner,
+            max_retries,
+            retry_backoff_factor,
+        }
+    }
+
+    pub fn get<U: IntoUrl>(&self, url: U) -> RequestBuilder {
+        self.request(Method::GET, url)
+    }
+
+    pub fn post<U: IntoUrl>(&self, url: U) -> RequestBuilder {
+        self.request(Method::POST, url)
+    }
+
+    pub fn put<U: IntoUrl>(&self, url: U) -> RequestBuilder {
+        self.request(Method::PUT, url)
+    }
+
+    pub fn patch<U: IntoUrl>(&self, url: U) -> RequestBuilder {
+        self.request(Method::PATCH, url)
+    }
+
+    pub fn delete<U: IntoUrl>(&self, url: U) -> RequestBuilder {
+        self.request(Method::DELETE, url)
+    }
+
+    pub fn head<U: IntoUrl>(&self, url: U) -> RequestBuilder {
+        self.request(Method::HEAD, url)
+    }
+
+    pub fn request<U: IntoUrl>(&self, method: Method, url: U) -> RequestBuilder {
+        self.inner.request(method, url)
+    }
+
+    pub fn execute(&self, request: Request) -> reqwest::Result<Response> {
+        let mut i: i32 = 0;
+
+        loop {
+            let duplicate_request = request.try_clone().unwrap();
+            let result = self.inner.execute(duplicate_request);
+
+            if let Ok(res) = &result {
+                match res.status().as_u16() {
+                    429 | 409 | 500..600 => {}
+                    _ => return result,
+                }
+            }
+
+            if i == self.max_retries {
+                return result;
+            }
+
+            let backoff = self.retry_backoff_factor * 2_f32.powi(i);
```

Comment:
shouldn't it be `self.retry_backoff_factor.powi(i)` ?

Reply:
I understand it's the same formula as the Python backoff, but it gives less control as having `initial_interval*factor.powi(i)`